### PR TITLE
Fix errors reported by gcc-8.1.0.

### LINKED ISSUE #16: Internal compiler error with Intel 17?
The following "bad pointer" error is obtained when attempting to compile mdspan with Intel 17.0.4:
```
experimental/__p0009_bits/basic_mdspan.hpp(124): internal error: bad pointer
    MDSPAN_TEMPLATE_REQUIRES(
    ^
```

Line 124 corresponds to the following code:
```
  MDSPAN_TEMPLATE_REQUIRES(
    class... IndexType,
    /* requires */ (
      _MDSPAN_FOLD_AND(_MDSPAN_TRAIT(is_convertible, IndexType, index_type) /* && ... */) &&
      (sizeof...(IndexType) == extents_type::rank_dynamic()) &&
      _MDSPAN_TRAIT(is_constructible, mapping_type, extents_type) &&
      _MDSPAN_TRAIT(is_default_constructible, accessor_type)
    )
  )
```

The compiler flags are:
```
C FLAGS: -w1 -vec-report0 -diag-disable remark -shared-intel -no-ftz -fma -diag-disable 11060 
     -xHost -qopenmp -g -O0 -inline-level=0 -ftrapuv -check=uninit -fp-model precise -fp-speculation 
     -debug inline-debug-info -fno-omit-frame-pointer -DDEBUG 
CXX FLAGS: -w1 -vec-report0 -diag-disable remark -shared-intel -no-ftz -fma -diag-disable 11060 
     -qopenmp -g -O0 -inline-level=0 -ftrapuv -check=uninit -fp-model precise -fp-speculation safe 
     inline-debug-info -fno-omit-frame-pointer -DDEBUG -early-template-check
```

Any information on a fix or workaround is much appreciated.

### LINKED ISSUE #20: gcc-8.1 compilation issue after #19 merged
After #19 was merged, we are seeing the following gcc-8.1 compile error:

```console
In file included from /home/kellyt/draco/src/experimental/mdspan:72,
                 from /home/kellyt/draco/src/experimental/test/tst_mdspan.cc:12:
/home/kellyt/draco/src/experimental/__p0009_bits/layout_stride.hpp:206:1: error: embedding a directive within macro arguments is not portable [-Werror]
 #if defined(__INTEL_COMPILER)
 ^
/home/kellyt/draco/src/experimental/__p0009_bits/layout_stride.hpp:209:1: error: embedding a directive within macro arguments is not portable [-Werror]
 #else
 ^
/home/kellyt/draco/src/experimental/__p0009_bits/layout_stride.hpp:211:1: error: embedding a directive within macro arguments is not portable [-Werror]
 #endif
 ^
```

* On a positive note, #19 resolved a fatal build error when we attempt to use intel-17.0.4.

cc: @brryan

### PR DESCRIPTION
**Pull request by KineticTheory**
* Fixes #20
* Fixes #16 

* This PR changed a couple of `#if MACRO` to `#ifdef MACRO` to silence gcc warnings about processing undefined macros.
* In `layout_stride.hpp`, I rewrote come logic to eliminate the embedded CPP macro reported in #20.

cc: @brryan

### CODE REVIEW

**Comment by KineticTheory** [comment]
@dhollman Can you review this small change to your #19?  This allows us to use `master` w/o any adjustments.  Thank you!

File: include/experimental/__p0009_bits/layout_stride.hpp [hunk 1]
```diff
@@ -198,19 +198,27 @@ class layout_stride_impl<
...
+    layout_stride_impl, (std::experimental::extents<Exts...> const& e), noexcept,
+    /* requires */ (
+      // remember that this also covers the zero strides case because an && fold on an empty pack is true
       _MDSPAN_FOLD_AND(Strides != dynamic_extent /* && ... */)
-#endif
     )
-  ) : __base_t(__base_t{__member_pair_t(e, __strides_storage_t())})
+  )
+#endif
+  : __base_t(__base_t{__member_pair_t(e, __strides_storage_t())})
```
Comment (dhollman) [inline include/experimental/__p0009_bits/layout_stride.hpp:221]: I'm not sure what this change accomplishes other than a stylistic change. Can you summarize?

File: include/experimental/__p0009_bits/macros.hpp [hunk 2]
```diff
@@ -211,7 +211,7 @@
 //==============================================================================
 // <editor-fold desc="inline variables"> {{{1
 
-#if _MDSPAN_USE_INLINE_VARIABLES
+#ifdef _MDSPAN_USE_INLINE_VARIABLES
```
Comment (dhollman) [inline include/experimental/__p0009_bits/macros.hpp:214]: If we really want to be safe on this we should do:

```suggestion
#if defined(_MDSPAN_USE_INLINE_VARIABLES) && _MDSPAN_USE_INLINE_VARIABLES != 0
```

File: include/experimental/__p0009_bits/layout_stride.hpp (hunk 1 above)
Comment (dhollman) [reply include/experimental/__p0009_bits/layout_stride.hpp:221]: Never mind; I finally got a hold of a copy of Visual Studio and this makes a lot more sense now.
